Long stack trace printed when test name is invalid

When the rgt.inp file specifies an undefined test (for example, if you developed a new test on another machine and forgot to push the changes), the harness currently prints a long and unhelpful stack trace:
```
Traceback (most recent call last):
  File "/sw/acceptance/olcf-test-harness/harness/bin/test_harness_driver.py", line 524, in <module>
    test_harness_driver()
  File "/sw/acceptance/olcf-test-harness/harness/bin/test_harness_driver.py", line 469, in test_harness_driver
    jstatus.initialize_subtest(launch_id, unique_id)
  File "/sw/acceptance/olcf-test-harness/harness/libraries/status_file.py", line 305, in initialize_subtest
    event_time = self.log_event(StatusFile.EVENT_LOGGING_START)
  File "/sw/acceptance/olcf-test-harness/harness/libraries/status_file.py", line 348, in log_event
    event_type, event_subtype, str(event_value))
  File "/sw/acceptance/olcf-test-harness/harness/libraries/status_file.py", line 512, in __log_event
    event_time, event_filename)
  File "/sw/acceptance/olcf-test-harness/harness/libraries/status_file.py", line 688, in get_status_info
    'harness function being executed from wrong directory.')
AssertionError: harness function being executed from wrong directory.
```

--- a/harness/bin/runtests.py
+++ b/harness/bin/runtests.py
@@ -402,6 +402,11 @@ def runtests(my_arg_string=None):
     ifile = input_files.rgt_input_file(inputfilename=harness_arguments.inputfile,
                                        runmodecmd=harness_arguments.runmode)
     main_logger.info("Completed reading the harness input file.")
+
+    # Check if there were any tests found:
+    if len(ifile.get_tests()) == 0:
+        main_logger.error("No tests found in input file. Aborting.")
+        return
     
     main_logger.info("Reading the harness config file.")
     config = rgt_config_file(configfilename=harness_arguments.configfile)

--- a/harness/libraries/apptest.py
+++ b/harness/libraries/apptest.py
@@ -70,12 +70,13 @@ class subtest(base_apptest, apptest_layout):
                               local_path_to_tests,
                               tag)
 
+        print(f"Initializing with logger={logger}")
         apptest_layout.__init__(self,
                                 local_path_to_tests,
                                 name_of_application,
                                 name_of_subtest,
-                                tag,
-                                logger=logger)
+                                logger=logger,
+                                harness_id=tag)
 
         # Format of data is [<local_path_to_tests>, <application>, <test>]
         self.__apps_test_checked_out = []

--- a/harness/libraries/apptest.py
+++ b/harness/libraries/apptest.py
@@ -74,7 +74,8 @@ class subtest(base_apptest, apptest_layout):
                                 local_path_to_tests,
                                 name_of_application,
                                 name_of_subtest,
-                                tag)
+                                tag,
+                                logger=logger)
 
         # Format of data is [<local_path_to_tests>, <application>, <test>]
         self.__apps_test_checked_out = []

--- a/harness/libraries/apptest.py
+++ b/harness/libraries/apptest.py
@@ -70,7 +70,6 @@ class subtest(base_apptest, apptest_layout):
                               local_path_to_tests,
                               tag)
 
-        print(f"Initializing with logger={logger}")
         apptest_layout.__init__(self,
                                 local_path_to_tests,
                                 name_of_application,

--- a/harness/libraries/input_files.py
+++ b/harness/libraries/input_files.py
@@ -32,8 +32,13 @@ class rgt_input_file:
         self.__path_to_tests = ""
         self.__inputFileName = inputfilename
 
-        # Read the input file.
-        self.__read_file()
+        # Read the input file. Returns True upon successful read.
+        # If read_file fails, self.__tests is emptied and False is returned
+        err = self.__read_file()
+        if not err:
+            print("ERROR: Failed to parse input file.")
+            # Short-circuit upon failure
+            return
 
         # If a CLI task was input use that instead
         if USE_HARNESS_TASKS_IN_RGT_INPUT_FILE not in runmodecmd :
@@ -73,7 +78,7 @@ class rgt_input_file:
 
         for tmpline in lines:
 
-            #If this is a comment line, the continue to next line.
+            #If this is a comment line, then continue to next line.
             if self.__is_comment_line(tmpline):
                 continue
 
@@ -88,26 +93,33 @@ class rgt_input_file:
 
             # Parse the line,depending upon what type of entry it is.
             if firstword == rgt_input_file.test_entry:
-                # Determine the number of words. If the number of words is 3
+                # Check that there are at either 4 or 5 items in the line
+                if not (len(words) == 4 or len(words) == 5):
+                    log_message = "Invalid number of words in test line: " + tmpline
+                    print(log_message)
+                    # Clear all tests -- invalid line in input file
+                    self.__tests = []
+                    return False
+                # Determine the number of words. If the number of words is 4
                 # The we run an indefinite number of times. If the number
-                # of words is 4, the we run a definite number of times dictated
-                # by the third word.
+                # of words is 5, the we run a definite number of times dictated
+                # by the last word.
                 app = words[2]
                 subtest = words[3]
                 nm_iters = -1
                 if len(words) == 5:
                     nm_iters = int(words[4])
-                elif len(words) > 5:
-                    log_message = "Invalid number of words in test line: " + tmpline
-                    print(log_message)
                 self.__tests.append([app,subtest,nm_iters])
 
             elif firstword == rgt_input_file.path_to_test_entry:
                 if (len(words) == 3):
+                    # Validate Path_to_tests here:
                     self.__path_to_tests = words[2]
-                elif len(words) > 3:
+                else:
                     log_message = "Invalid number of words in path line: " + tmpline
                     print(log_message)
+                    self.__tests = []
+                    return False
 
             elif firstword == rgt_input_file.harness_task_entry:
                 if (len(words) == 3):
@@ -117,9 +129,14 @@ class rgt_input_file:
                 else:
                     log_message = "Invalid number of words in task line: " + tmpline
                     print(log_message)
+                    self.__tests = []
+                    return False
             else:
                 log_message = "Invalid line: " + tmpline
                 print(log_message)
+                self.__tests = []
+                return False
+        return True
 
     def __is_comment_line(self,word):
         if word[0] == rgt_input_file.comment_line_entry:

--- a/harness/libraries/layout_of_apps_directory.py
+++ b/harness/libraries/layout_of_apps_directory.py
@@ -130,13 +130,13 @@ class apptest_layout:
     # Debug function.
     #
     def debug_layout(self):
-        print ("\n\n")
-        print ("================================================================")
-        print ("Debugging local layout " + self.__appname + self.__testname)
-        print ("================================================================")
+        self.__logger.doDebugLogging ("\n\n")
+        self.__logger.doDebugLogging ("================================================================")
+        self.__logger.doDebugLogging ("Debugging local layout " + self.__appname + self.__testname)
+        self.__logger.doDebugLogging ("================================================================")
         for key in self.__apptest_layout.keys():
-            print ("%-20s = %-20s" % (key, self.__apptest_layout[key]))
-        print ("================================================================\n\n")
+            self.__logger.doDebugLogging ("%-20s = %-20s" % (key, self.__apptest_layout[key]))
+        self.__logger.doDebugLogging ("================================================================\n\n")
 
     #
     # Returns the path to the application directory.

--- a/harness/libraries/layout_of_apps_directory.py
+++ b/harness/libraries/layout_of_apps_directory.py
@@ -109,9 +109,17 @@ class apptest_layout:
     # Check that the required paths to source and scripts exist
     def check_paths(self):
         """ Returns False if the Source dir, Scripts dir, or test input ini files don't exist """
+        # Check that the Application dir exists
+        if not os.path.exists(self.__apptest_layout['app']):
+            self.__logger.doErrorLogging(f"Could not find the Application root directory for App={self.__appname}, Test={self.__testname}.")
+            return False
         # Check that the Application's Source dir exists
         if not os.path.exists(self.get_path_to_source()):
             self.__logger.doErrorLogging(f"Could not find the Source directory for App={self.__appname}, Test={self.__testname}.")
+            return False
+        # Check that the Test dir exists
+        if not os.path.exists(self.__apptest_layout['test']):
+            self.__logger.doErrorLogging(f"Could not find the test directory for App={self.__appname}, Test={self.__testname}.")
             return False
         # Check that the Scripts directory exists
         if not os.path.exists(self.get_path_to_scripts()):

--- a/harness/libraries/layout_of_apps_directory.py
+++ b/harness/libraries/layout_of_apps_directory.py
@@ -106,6 +106,23 @@ class apptest_layout:
         self.__apptest_layout = copy.deepcopy(apptest_layout.directory_structure_template)
         self.__setApplicationTestLayout()
 
+    # Check that the required paths to source and scripts exist
+    def check_paths(self):
+        """ Returns False if the Source dir, Scripts dir, or test input ini files don't exist """
+        # Check that the Application's Source dir exists
+        if not os.path.exists(self.get_path_to_source()):
+            self.__logger.doErrorLogging(f"Could not find the Source directory for App={self.__appname}, Test={self.__testname}.")
+            return False
+        # Check that the Scripts directory exists
+        if not os.path.exists(self.get_path_to_scripts()):
+            self.__logger.doErrorLogging(f"Could not find the Scripts directory for App={self.__appname}, Test={self.__testname}.")
+            return False
+        # Check that the an rgt_test_ini file exists (either .ini or .txt)
+        if not (os.path.exists(self.__apptest_layout['test_input_ini']) or os.path.exists(self.__apptest_layout['test_input_txt'])):
+            self.__logger.doErrorLogging(f"Could not find the test input file for App={self.__appname}, Test={self.__testname}.")
+            return False
+        return True
+
     # Return the harness id for the test (may be None)
     def get_harness_id(self):
         return self.__testid

--- a/harness/libraries/layout_of_apps_directory.py
+++ b/harness/libraries/layout_of_apps_directory.py
@@ -90,10 +90,12 @@ class apptest_layout:
                  applications_rootdir,
                  name_of_application,
                  name_of_subtest,
+                 logger=None,
                  harness_id=None):
         self.__applications_root = applications_rootdir
         self.__appname = name_of_application
         self.__testname = name_of_subtest
+        self.__logger = logger
         self.__workspace = None
 
         if harness_id == None:
@@ -211,11 +213,11 @@ class apptest_layout:
         #
         apptest_dir = self.get_path_to_test()
         latest_lnk = os.path.join(apptest_dir, apptest_layout.test_status_dirname, 'latest')
-        if os.path.exists(latest_lnk):
-            try:
-                os.unlink(latest_lnk)
-            except FileNotFoundError as e:
-                print("Found potential race condition while trying to change latest link. Skipping")
+        try:
+            # Not guarded by a conditional, so that it removes broken links
+            os.unlink(latest_lnk)
+        except FileNotFoundError as e:
+            self.__logger.doInfoLogging("Could not remove 'latest' link in Status.")
         try_symlink(spath, latest_lnk)
 
         return spath
@@ -238,11 +240,11 @@ class apptest_layout:
         #
         apptest_dir = self.get_path_to_test()
         latest_lnk = os.path.join(apptest_dir, apptest_layout.test_run_archive_dirname, 'latest')
-        if os.path.exists(latest_lnk):
-            try:
-                os.unlink(latest_lnk)
-            except FileNotFoundError as e:
-                print("Found potential race condition while trying to change latest link. Skipping")
+        try:
+            # Not guarded by a conditional, so that it removes broken links
+            os.unlink(latest_lnk)
+        except FileNotFoundError as e:
+            self.__logger.doInfoLogging("Could not remove 'latest' link in Run_Archive.")
         try_symlink(rpath, latest_lnk)
 
         return rpath

--- a/harness/libraries/regression_test.py
+++ b/harness/libraries/regression_test.py
@@ -79,6 +79,8 @@ class Harness:
         if testshot_key in testshot_cfg.keys():
             testshot_str = testshot_cfg[testshot_key]
         self.__launch_id = f'{testshot_str}/{user_str}@{time_str}'
+        self.__launched_tests = 0
+        self.__skipped_tests = 0
 
         # Define a logger that streams to file.
         logger_name=Harness.LOGGER_NAME
@@ -250,7 +252,16 @@ class Harness:
                                                       logger = a_logger,
                                                       tag=self.__timestamp)
 
-                app_subtests[appname].append(subtest)
+                if subtest.check_paths():
+                    app_subtests[appname].append(subtest)
+                    self.__launched_tests += 1
+                else:
+                    # Then at least 1 required path (ie: source, scripts directories) does not exist
+                    # Exact error messages are printed by check_paths
+                    a_logger.doWarningLogging(f"Skipping App={appname}, Test={testname}.")
+                    # Log to both loggers
+                    self.__myLogger.doWarningLogging(f"Skipping App={appname}, Test={testname}.")
+                    self.__skipped_tests += 1
 
         return app_subtests
 
@@ -283,6 +294,8 @@ class Harness:
 
             message = "All applications completed futures. Yahoo!!"
             self.__myLogger.doInfoLogging(message)
+            # For the moment, hard-code this as a print statement.
+            print(f"Skipped {self.__skipped_tests}, launched {self.__launched_tests}.")
 
         return
 

